designs/messaging/messaging-framework.md

```diff
@@ -0,0 +1,413 @@
+
+- Mapping .NET objects to which topics, queues or event bus to publish to (Publish points)
+- Mapping .NET objects to handlers to process messages (Subscriber points)
+- Concurrency and message pump settings
+
+#### Configuration mapping
+For both publishing and subscribing there are 3 important pieces of information to configure.
+
+- The .NET type to use as the container of the message
+- The AWS resource identifier (queue URLs, topic ARNs, event bus ARNs)
+- The message type that will be set in the message envelope when creating the message. If this is not set then the full .NET type name will be used.
+
+A .NET type can only be registered once for publishing or subscribing. When a .NET object is being published the AWS resource and message type name will be retrieved from the mapping based on the .NET type. The .NET object will be serialized along with its message envelope.
+
+When reading messages from queues the message type will be pulled from the message envelope. The message type will be used to deserialize the message to the corresponding .NET type. The handler class that has been mapped to the .NET type will be invoked, passing in the .NET object with its message envelope for handlers that need extra metadata.
+
+#### Configuring a publisher
+An ASP.NET Core application could be a publisher. A typical ASP.NET Core application has a `ConfigureServices` method to configure the DI. In this method the application would configure the types of messages it will send. The example below shows how to register .NET objects to publishing endpoints.
+
+```csharp
+public IConfiguration Configuration { get; }
+
+public void ConfigureServices(IServiceCollection services)
+{
+    // Pull out physical AWS resource ids from .NET Configuration framework
+    var awsResourceIds = new AWSResourceIds();
+    Configuration.GetSection(AWSResourceIds.SectionName).Bind(awsResourceIds);
+            
+    services.AddAWSMessageBus(builder =>
+    {
+        builder.AddPublishingQueue<OrderInfo>(awsResourceIds.OrderInfoQueue);
+        
+        // By default the .NET object type name will be used as the message type name
+        // set in the message envelope. An optional name can be specified.
+        // This is useful when using multiple languages to have an language-agnostic name.
+        builder.AddPublishingQueue<OrderInfo>(awsResourceIds.OrderInfoQueue, "OrderMessageType");
+        
+        builder.AddPublishingTopic<HealthStatus>(awsResourceIds.HealthStatusTopic);
+        
+        // One could also specify PublishMetadata that will be set in the 
+        // message envelope for ALL messages published by this application
+        builder.AddPublishingEventBus<InventoryCheck>(awsResourceIds.WarehouseEventBus, new PublishMetadata
+        {
+            MessageType = "InventoryCheckMessage",
+            Source = "OrderHandlerSystem" 
```

Comment:
Would it also be possible to set metadata on the message bus as a whole? I can imagine that the above example of `Source` would naturally apply to all publishers in this assembly?